`NullPointerException` when using 0.9.12 and later to run tests with native image compilation

**Describe the bug**
Running a simple test with `native-maven-plugin` 0.9.12 and later is resulting a `NullPointerException` at build time. 

**To Reproduce**
Reproduce: https://github.com/mpeddada1/junit-testrule

**Expected behavior**
The tests run successfully with the versions in question.

**Logs**
Following the steps in the README of the reproducer is resulting in this stacktrace:

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running com.example.MySampleTest
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.028 s - in com.example.MySampleTest
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- native-maven-plugin:0.9.13:test (test-native) @ child-project ---
[INFO] ====================
[INFO] Initializing project: child-project
[INFO] ====================
[INFO] Found GraalVM installation from JAVA_HOME variable.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.770 s
[INFO] Finished at: 2022-07-20T17:43:43-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.graalvm.buildtools:native-maven-plugin:0.9.13:test (test-native) on project child-project: Execution test-native of goal org.graalvm.buildtools:native-maven-plugin:0.9.13:test failed.: NullPointerException -> [Help 1]
```
[Snippet of stacktrace with -X flag ](https://gist.github.com/mpeddada1/9b145382933c837489ec8d1aa56e3cc9)

However, **note** that this test runs successfully with `0.9.11`

**System Info (please complete the following information):**
 - OS: `Linux`
 - GraalVM Version: 22.1.0 CE
 - Java Version: 11
 - Plugin version: native-maven-plugin 0.9.12 and 0.9.13

I'm wondering if anyone has run into this problem? Thank you!

cc @suztomo @meltsufin

Nice. It shows the exact line number that hit NullPointerException. Let's wait for the maintainers.

```
Caused by: java.lang.NullPointerException
    at org.graalvm.buildtools.maven.AbstractNativeMojo.processArtifact (AbstractNativeMojo.java:285)
    at org.graalvm.buildtools.maven.AbstractNativeMojo.getMainBuildPath (AbstractNativeMojo.java:351)
    at org.graalvm.buildtools.maven.AbstractNativeMojo.populateApplicationClasspath (AbstractNativeMojo.java:361)
    at org.graalvm.buildtools.maven.NativeTestMojo.populateApplicationClasspath (NativeTestMojo.java:88)
    at org.graalvm.buildtools.maven.AbstractNativeMojo.populateClasspath (AbstractNativeMojo.java:372)
    at org.graalvm.buildtools.maven.AbstractNativeMojo.getClasspath (AbstractNativeMojo.java:378)
    at org.graalvm.buildtools.maven.AbstractNativeMojo.getBuildArgs (AbstractNativeMojo.java:211)
    at org.graalvm.buildtools.maven.AbstractNativeMojo.buildImage (AbstractNativeMojo.java:393)
    at org.graalvm.buildtools.maven.NativeTestMojo.execute (NativeTestMojo.java:149)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)

```

--- a/common/utils/src/main/java/org/graalvm/buildtools/utils/FileUtils.java
+++ b/common/utils/src/main/java/org/graalvm/buildtools/utils/FileUtils.java
@@ -52,8 +52,10 @@ import java.nio.file.Path;
 import java.nio.file.StandardCopyOption;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
 import java.util.Optional;
 import java.util.function.Consumer;
+import java.util.stream.Stream;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipInputStream;
 

--- a/common/utils/src/main/java/org/graalvm/buildtools/utils/NativeImageUtils.java
+++ b/common/utils/src/main/java/org/graalvm/buildtools/utils/NativeImageUtils.java
@@ -45,6 +45,7 @@ import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.nio.file.StandardOpenOption;
 import java.util.Collections;
 import java.util.List;
@@ -72,15 +73,24 @@ public class NativeImageUtils {
         return "native-image-configure" + GRAALVM_EXE_EXTENSION;
     }
 
-    public static List<String> convertToArgsFile(List<String> cliArgs) {
+    public static List<String> convertToArgsFile(List<String> cliArgs, Path outputDir) {
+        return convertToArgsFile(cliArgs, outputDir, Paths.get(""));
+    }
+
+    public static List<String> convertToArgsFile(List<String> cliArgs, Path outputDir, Path projectDir) {
         try {
-            File tmpFile = Files.createTempFile("native-image", "args").toFile();
-            tmpFile.deleteOnExit();
+            boolean ignored = outputDir.toFile().mkdirs();
+            File tmpFile = Files.createTempFile(outputDir, "native-image-", ".args").toFile();
+            // tmpFile.deleteOnExit();
             cliArgs = cliArgs.stream().map(NativeImageUtils::escapeArg).collect(Collectors.toList());
             Files.write(tmpFile.toPath(), cliArgs, StandardCharsets.UTF_8, StandardOpenOption.CREATE);
-            return Collections.singletonList("@" + tmpFile.getAbsolutePath());
-        } catch (IOException e) {
 
+            Path resultingPath = tmpFile.toPath().toAbsolutePath();
+            if (projectDir != null) { // We know where the project dir is, so want to use relative paths
+                resultingPath = projectDir.toAbsolutePath().relativize(resultingPath);
+            }
+            return Collections.singletonList("@" + resultingPath);
+        } catch (IOException e) {
             return Collections.unmodifiableList(cliArgs);
         }
     }

--- a/docs/src/docs/asciidoc/index.adoc
+++ b/docs/src/docs/asciidoc/index.adoc
@@ -16,17 +16,26 @@ If you are using alternative build systems, see <<alternative-build-systems.adoc
 
 [[changelog]]
 == Changelog
+
 === Release 0.9.14
 
-=== Gradle plugin
+==== Gradle plugin
 * Add ability to set environment variables to the native image builder process
+* Argument files are now stored in the `build` directory (workaround for absolute path issue on Windows with older GraalVM versions)
+
+==== Maven plugin
+* Added `native:compile` forking goal that can be started from the command line as `mvn native:compile`.
+* Argument files are now stored in the `target` directory (workaround for absolute path issue on Windows with older GraalVM versions).
+* Default and test outputs are now much less noisy.
+* When running tests in JVM mode with the native-image-agent, GraalVM's `java` executable is now always used.
+* Maven plugin now shouldn't require that JVM running it must be GraalVM.
 
 === Release 0.9.13
 
-=== Gradle plugin
+==== Gradle plugin
 * Reverted a change in the `NativeImagePlugin` that removed publicly accessible constants. This should prevent breakage of external plugins.
 
-=== JUnit testing support
+==== JUnit testing support
 * Adapted the JUnit automatic metadata registration to changes in annotation handling on newer native image versions.
 
 === Release 0.9.12

--- a/docs/src/docs/asciidoc/maven-plugin.adoc
+++ b/docs/src/docs/asciidoc/maven-plugin.adoc
@@ -70,6 +70,8 @@ If the heuristics fail with the `no main manifest attribute, in target/<name>.ja
 specified in the `<configuration>` node of the plugin.
 When `mvn -Pnative package` completes, an executable is ready for use, generated in the `target` directory of the project.
 
+NOTE: Building native image without attaching to the `package` phase can be done by invoking the forking `compile` goal directly, e.g. `mvn native:compile`
+
 [TIP]
 .Testing pre-releases
 ====

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/NativeImageCommandLineProvider.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/NativeImageCommandLineProvider.java
@@ -53,6 +53,8 @@ import org.gradle.api.tasks.Nested;
 import org.gradle.process.CommandLineArgumentProvider;
 
 import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
@@ -64,16 +66,19 @@ public class NativeImageCommandLineProvider implements CommandLineArgumentProvid
     private final Provider<NativeImageOptions> options;
     private final Provider<String> executableName;
     private final Provider<String> outputDirectory;
+    private final Provider<String> workingDirectory;
     private final Provider<RegularFile> classpathJar;
     private final Provider<Boolean> useArgFile;
 
     public NativeImageCommandLineProvider(Provider<NativeImageOptions> options,
                                           Provider<String> executableName,
+                                          Provider<String> workingDirectory,
                                           Provider<String> outputDirectory,
                                           Provider<RegularFile> classpathJar,
                                           Provider<Boolean> useArgFile) {
         this.options = options;
         this.executableName = executableName;
+        this.workingDirectory = workingDirectory;
         this.outputDirectory = outputDirectory;
         this.classpathJar = classpathJar;
         this.useArgFile = useArgFile;
@@ -142,7 +147,8 @@ public class NativeImageCommandLineProvider implements CommandLineArgumentProvid
         }
         cliArgs.addAll(options.getBuildArgs().get());
         if (useArgFile.getOrElse(true)) {
-            return NativeImageUtils.convertToArgsFile(cliArgs);
+            Path argFileDir = Paths.get(workingDirectory.get());
+            return NativeImageUtils.convertToArgsFile(cliArgs, argFileDir, argFileDir);
         }
         return Collections.unmodifiableList(cliArgs);
 

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/BuildNativeImageTask.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/BuildNativeImageTask.java
@@ -152,6 +152,7 @@ public abstract class BuildNativeImageTask extends DefaultTask {
         return new NativeImageCommandLineProvider(
                 getOptions(),
                 getExecutableShortName(),
+                getProviders().provider(() -> getWorkingDirectory().get().getAsFile().getAbsolutePath()),
                 // Can't use getOutputDirectory().map(...) because Gradle would complain that we use
                 // a mapped value before the task was called, when we are actually calling it...
                 getProviders().provider(() -> getOutputDirectory().getAsFile().get().getAbsolutePath()),

--- a/native-maven-plugin/build-plugins/src/main/java/org/graalvm/build/maven/MavenTask.java
+++ b/native-maven-plugin/build-plugins/src/main/java/org/graalvm/build/maven/MavenTask.java
@@ -90,12 +90,12 @@ public abstract class MavenTask extends DefaultTask {
             spec.setClasspath(getMavenEmbedderClasspath());
             spec.getMainClass().set("org.apache.maven.cli.MavenCli");
             spec.systemProperty("maven.multiModuleProjectDirectory", projectdir.getAbsolutePath());
+            spec.systemProperty("org.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener", "warn");
             prepareSpec(spec);
             List<String> arguments = new ArrayList<>();
             arguments.addAll(Arrays.asList(
                     "--errors",
                     "-U",
-                    "-Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn",
                     "--batch-mode",
                     "--settings", settingsFile.getAbsolutePath(),
                     "--file", pomFile.getAbsolutePath()

--- a/native-maven-plugin/build.gradle.kts
+++ b/native-maven-plugin/build.gradle.kts
@@ -63,6 +63,7 @@ dependencies {
     implementation(libs.utils)
     implementation(libs.jackson.databind)
     implementation(libs.jvmReachabilityMetadata)
+    implementation(libs.graalvm.svm)
 
     compileOnly(libs.maven.pluginApi)
     compileOnly(libs.maven.core)

--- a/native-maven-plugin/build.gradle.kts
+++ b/native-maven-plugin/build.gradle.kts
@@ -173,5 +173,5 @@ tasks {
 }
 
 tasks.withType<Checkstyle>().configureEach {
-    setConfigFile(layout.projectDirectory.dir("../config/checkstyle.xml").asFile)
+    configFile = layout.projectDirectory.dir("../config/checkstyle.xml").asFile
 }

--- a/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationFunctionalTest.groovy
+++ b/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationFunctionalTest.groovy
@@ -100,7 +100,7 @@ class JavaApplicationFunctionalTest extends AbstractGraalVMMavenFunctionalTest {
         withSample("java-application-with-custom-packaging")
 
         when:
-        mvn 'package', '-Dpackaging=native-image', 'exec:exec@native'
+        mvnDebug 'package', '-Dpackaging=native-image', 'exec:exec@native'
 
         then:
         buildSucceeded

--- a/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationWithAgentFunctionalTest.groovy
+++ b/native-maven-plugin/src/functionalTest/groovy/org/graalvm/buildtools/maven/JavaApplicationWithAgentFunctionalTest.groovy
@@ -54,7 +54,7 @@ class JavaApplicationWithAgentFunctionalTest extends AbstractGraalVMMavenFunctio
         when:
         // Run Maven in debug mode (-X) in order to capture the command line arguments
         // used to launch Surefire with the agent.
-        mvn '-X', '-Pnative', 'test', '-DskipNativeTests'
+        mvnDebug '-Pnative', 'test', '-DskipNativeTests'
 
         then:
         // Agent is used with Surefire
@@ -85,7 +85,7 @@ class JavaApplicationWithAgentFunctionalTest extends AbstractGraalVMMavenFunctio
         when:
         // Run Maven in debug mode (-X) in order to capture the command line arguments
         // used to launch Surefire with the agent.
-        mvn '-X', '-Pnative', 'test'
+        mvnDebug '-Pnative', 'test'
 
         then:
         outputContains """
@@ -162,7 +162,7 @@ class JavaApplicationWithAgentFunctionalTest extends AbstractGraalVMMavenFunctio
         when:
         // Run Maven in debug mode (-X) in order to capture the command line arguments
         // used to launch Surefire with the agent.
-        mvn '-X', '-Pnative', '-DagentOptions=periodic-config', 'test'
+        mvnDebug '-Pnative', '-DagentOptions=periodic-config', 'test'
 
         then:
         outputContains """
@@ -214,7 +214,7 @@ class JavaApplicationWithAgentFunctionalTest extends AbstractGraalVMMavenFunctio
         withSample("java-application-with-reflection")
 
         when:
-        mvn '-X', '-Pnative', '-DskipTests=true', '-DskipNativeBuild=true', 'package', 'exec:exec@java-agent'
+        mvnDebug '-Pnative', '-DskipTests=true', '-DskipNativeBuild=true', 'package', 'exec:exec@java-agent'
 
         then:
         ['jni', 'proxy', 'reflect', 'resource', 'serialization'].each { name ->
@@ -246,7 +246,7 @@ class JavaApplicationWithAgentFunctionalTest extends AbstractGraalVMMavenFunctio
         withSample("java-application-with-reflection")
 
         when:
-        mvn '-X', '-Pnative', '-Dagent=false', '-DskipTests=true', '-DskipNativeBuild=true', 'package', 'exec:exec@java-agent'
+        mvnDebug '-Pnative', '-Dagent=false', '-DskipTests=true', '-DskipNativeBuild=true', 'package', 'exec:exec@java-agent'
 
         then:
         outputDoesNotContain '-agentlib:native-image-agent'
@@ -263,7 +263,7 @@ class JavaApplicationWithAgentFunctionalTest extends AbstractGraalVMMavenFunctio
         withSample("java-application-with-reflection")
 
         when:
-        mvn '-X', '-Pnative', '-DagentOptions=periodic-config', '-DskipTests=true', '-DskipNativeBuild=true', 'package', 'exec:exec@java-agent'
+        mvnDebug '-Pnative', '-DagentOptions=periodic-config', '-DskipTests=true', '-DskipNativeBuild=true', 'package', 'exec:exec@java-agent'
 
         then:
         ['jni', 'proxy', 'reflect', 'resource', 'serialization'].each { name ->

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/AbstractNativeMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/AbstractNativeMojo.java
@@ -323,8 +323,9 @@ public abstract class AbstractNativeMojo extends AbstractMojo {
                         valid = valid && relativeSubDir.getName(0).toString().equals(artifact.getGroupId());
                         valid = valid && relativeSubDir.getName(1).toString().equals(artifact.getArtifactId());
                         if (!valid) {
-                            String example = NATIVE_IMAGE_META_INF + "/${groupId}/${artifactId}/" + NATIVE_IMAGE_PROPERTIES_FILENAME;
-                            logger.warn(nativeImageProperty.toUri() + " does not match recommended " + example + " layout.");
+                            String example = NATIVE_IMAGE_META_INF + "/%s/%s/" + NATIVE_IMAGE_PROPERTIES_FILENAME;
+                            example = String.format(example, artifact.getGroupId(), artifact.getArtifactId());
+                            logger.warn("Properties file at '" + nativeImageProperty.toUri() + "' does not match the recommended '" + example + "' layout.");
                         }
                     }
                 }

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/AbstractNativeMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/AbstractNativeMojo.java
@@ -271,7 +271,8 @@ public abstract class AbstractNativeMojo extends AbstractMojo {
         }
 
         if (useArgFile) {
-            return NativeImageUtils.convertToArgsFile(cliArgs);
+            Path tmpDir = Paths.get("target", "tmp");
+            return NativeImageUtils.convertToArgsFile(cliArgs, tmpDir);
         }
         return Collections.unmodifiableList(cliArgs);
     }
@@ -405,7 +406,6 @@ public abstract class AbstractNativeMojo extends AbstractMojo {
             if (!outputDirectory.exists() && !outputDirectory.mkdirs()) {
                 throw new MojoExecutionException("Failed creating output directory");
             }
-            processBuilder.directory(outputDirectory);
             processBuilder.inheritIO();
 
             String commandString = String.join(" ", processBuilder.command());

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/AbstractNativeMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/AbstractNativeMojo.java
@@ -278,6 +278,12 @@ public abstract class AbstractNativeMojo extends AbstractMojo {
 
     protected Path processArtifact(Artifact artifact, String artifactType) throws MojoExecutionException {
         File artifactFile = artifact.getFile();
+
+        if (artifactFile == null) {
+            logger.debug("Missing artifact file for artifact " + artifact + " (type: " + artifact.getType() + ")");
+            return null;
+        }
+
         if (!artifactType.equals(artifact.getType())) {
             logger.warn("Ignoring non-jar type ImageClasspath Entry " + artifact);
             return null;
@@ -288,7 +294,7 @@ public abstract class AbstractNativeMojo extends AbstractMojo {
         }
 
         Path jarFilePath = artifactFile.toPath();
-        logger.info("ImageClasspath Entry: " + artifact + " (" + jarFilePath.toUri() + ")");
+        logger.debug("ImageClasspath Entry: " + artifact + " (" + jarFilePath.toUri() + ")");
 
         warnIfWrongMetaInfLayout(jarFilePath, artifact);
         return jarFilePath;
@@ -300,7 +306,7 @@ public abstract class AbstractNativeMojo extends AbstractMojo {
 
     protected void warnIfWrongMetaInfLayout(Path jarFilePath, Artifact artifact) throws MojoExecutionException {
         if (jarFilePath.toFile().isDirectory()) {
-            logger.warn("Artifact `" + jarFilePath + "` is a directory.");
+            logger.debug("Artifact `" + jarFilePath + "` is a directory.");
             return;
         }
         URI jarFileURI = URI.create("jar:" + jarFilePath.toUri());

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeCompileMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeCompileMojo.java
@@ -1,0 +1,61 @@
+/*
+ * Copyright (c) 2020, 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * The Universal Permissive License (UPL), Version 1.0
+ *
+ * Subject to the condition set forth below, permission is hereby granted to any
+ * person obtaining a copy of this software, associated documentation and/or
+ * data (collectively the "Software"), free of charge and under any and all
+ * copyright rights in the Software, and any and all patent rights owned or
+ * freely licensable by each licensor hereunder covering either (i) the
+ * unmodified Software as contributed to or provided by such licensor, or (ii)
+ * the Larger Works (as defined below), to deal in both
+ *
+ * (a) the Software, and
+ *
+ * (b) any piece of software and/or hardware listed in the lrgrwrks.txt file if
+ * one is included with the Software each a "Larger Work" to which the Software
+ * is contributed by such licensors),
+ *
+ * without restriction, including without limitation the rights to copy, create
+ * derivative works of, display, perform, and distribute the Software and make,
+ * use, sell, offer for sale, import, export, have made, and have sold the
+ * Software and the Larger Work(s), and to sublicense the foregoing rights on
+ * either these or other terms.
+ *
+ * This license is subject to the following condition:
+ *
+ * The above copyright notice and either this complete permission notice or at a
+ * minimum a reference to the UPL must be included in all copies or substantial
+ * portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package org.graalvm.buildtools.maven;
+
+import org.apache.maven.plugins.annotations.Execute;
+import org.apache.maven.plugins.annotations.LifecyclePhase;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.ResolutionScope;
+
+
+/**
+ * This goal builds native images.
+ * It can be invoked using `mvn native:compile`.
+ */
+@Mojo(name = "compile", defaultPhase = LifecyclePhase.PACKAGE,
+        requiresDependencyResolution = ResolutionScope.RUNTIME,
+        requiresDependencyCollection = ResolutionScope.RUNTIME)
+@Execute(phase = LifecyclePhase.PACKAGE)
+@SuppressWarnings("unused")
+public class NativeCompileMojo extends NativeCompileNoForkMojo {
+    // no-op
+}

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeCompileMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeCompileMojo.java
@@ -49,7 +49,7 @@ import org.apache.maven.plugins.annotations.ResolutionScope;
 
 /**
  * This goal builds native images.
- * It can be invoked using `mvn native:compile`.
+ * It should be invoked from the command line as a single goal (`mvn native:compile`).
  */
 @Mojo(name = "compile", defaultPhase = LifecyclePhase.PACKAGE,
         requiresDependencyResolution = ResolutionScope.RUNTIME,
@@ -57,5 +57,4 @@ import org.apache.maven.plugins.annotations.ResolutionScope;
 @Execute(phase = LifecyclePhase.PACKAGE)
 @SuppressWarnings("unused")
 public class NativeCompileMojo extends NativeCompileNoForkMojo {
-    // no-op
 }

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeCompileNoForkMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeCompileNoForkMojo.java
@@ -58,10 +58,15 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.function.BiFunction;
 
+
+/**
+ * This goal runs native builds. It functions the same as the native:compile goal, but it
+ * does not fork the build, so it is suitable for attaching to the build lifecycle.
+ */
 @Mojo(name = "build", defaultPhase = LifecyclePhase.PACKAGE,
         requiresDependencyResolution = ResolutionScope.RUNTIME,
         requiresDependencyCollection = ResolutionScope.RUNTIME)
-public class NativeBuildMojo extends AbstractNativeMojo {
+public class NativeCompileNoForkMojo extends AbstractNativeMojo {
 
     @Parameter(property = "skipNativeBuild", defaultValue = "false")
     private boolean skip;
@@ -79,7 +84,7 @@ public class NativeBuildMojo extends AbstractNativeMojo {
     @Override
     public void execute() throws MojoExecutionException {
         if (skip) {
-            getLog().info("Skipping native-image generation (parameter 'skipNativeBuild' is true).");
+            logger.info("Skipping native-image generation (parameter 'skipNativeBuild' is true).");
             return;
         }
 
@@ -147,7 +152,7 @@ public class NativeBuildMojo extends AbstractNativeMojo {
             mainClass = mainClassProvider.apply(pluginName, nodeNames);
 
             if (mainClass != null) {
-                getLog().info("Obtained main class from plugin " + pluginName + " with the following path: " + String.join(" -> ", nodeNames));
+                logger.info("Obtained main class from plugin " + pluginName + " with the following path: " + String.join(" -> ", nodeNames));
             }
         }
     }

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeExtension.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeExtension.java
@@ -73,7 +73,7 @@ public class NativeExtension extends AbstractMavenLifecycleParticipant implement
     private static final String JUNIT_PLATFORM_LISTENERS_UID_TRACKING_OUTPUT_DIR = "junit.platform.listeners.uid.tracking.output.dir";
     private static final String NATIVEIMAGE_IMAGECODE = "org.graalvm.nativeimage.imagecode";
 
-    private Logger logger;
+    private static Logger logger;
 
     @Override
     public void enableLogging(Logger logger) {
@@ -91,6 +91,8 @@ public class NativeExtension extends AbstractMavenLifecycleParticipant implement
     static String testIdsDirectory(String baseDir) {
         return baseDir + File.separator + "test-ids";
     }
+
+    private static String graalvmJava;
 
     static String buildAgentArgument(String baseDir, Context context, List<String> agentOptions) {
         List<String> options = new ArrayList<>(agentOptions);
@@ -115,6 +117,12 @@ public class NativeExtension extends AbstractMavenLifecycleParticipant implement
 
     @Override
     public void afterProjectsRead(MavenSession session) {
+        try {
+            graalvmJava = Utils.getNativeImage(logger).getParent().resolve("java").toString();
+        } catch (MojoExecutionException e) {
+            throw new RuntimeException(e);
+        }
+
         for (MavenProject project : session.getProjects()) {
             Build build = project.getBuild();
             withPlugin(build, "native-maven-plugin", nativePlugin -> {
@@ -159,12 +167,7 @@ public class NativeExtension extends AbstractMavenLifecycleParticipant implement
                                     for (Xpp3Dom child : children) {
                                         commandlineArgs.addChild(child);
                                     }
-                                    Xpp3Dom executable = findOrAppend(config, "executable");
-                                    try {
-                                        executable.setValue(Utils.getNativeImage(logger).getParent().resolve("java").toString());
-                                    } catch (MojoExecutionException e) {
-                                        throw new RuntimeException(e);
-                                    }
+                                    findOrAppend(config, "executable").setValue(graalvmJava);
                                 }
                             })
                     );
@@ -318,6 +321,7 @@ public class NativeExtension extends AbstractMavenLifecycleParticipant implement
             Xpp3Dom argLine = new Xpp3Dom("argLine");
             argLine.setValue(agentArgument);
             configuration.addChild(argLine);
+            findOrAppend(configuration, "jvm").setValue(graalvmJava);
         });
     }
 

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeTestMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/NativeTestMojo.java
@@ -70,6 +70,7 @@ import java.util.stream.Stream;
 import static org.graalvm.buildtools.Utils.NATIVE_TESTS_EXE;
 
 /**
+ * This goal builds and runs native tests.
  * @author Sebastien Deleuze
  */
 @Mojo(name = "test", defaultPhase = LifecyclePhase.TEST, threadSafe = true,

--- a/native-maven-plugin/src/seeding-build/pom.xml
+++ b/native-maven-plugin/src/seeding-build/pom.xml
@@ -52,6 +52,7 @@
         <java.version>1.8</java.version>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <junit.jupiter.version>5.8.1</junit.jupiter.version>
+        <graalvm.version>22.0.0</graalvm.version>
         <mainClass>org.graalvm.demo.Application</mainClass>
     </properties>
 
@@ -61,6 +62,11 @@
             <artifactId>junit-jupiter</artifactId>
             <version>${junit.jupiter.version}</version>
             <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.graalvm.nativeimage</groupId>
+            <artifactId>svm</artifactId>
+            <version>${graalvm.version}</version>
         </dependency>
     </dependencies>
 

--- a/native-maven-plugin/src/seeding-build/pom.xml
+++ b/native-maven-plugin/src/seeding-build/pom.xml
@@ -80,6 +80,14 @@
                 <configuration>
                     <source>${java.version}</source>
                     <target>1.8</target>
+                </configuration>
+            </plugin>
+
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-jar-plugin</artifactId>
+                <version>3.2.2</version>
+                <configuration>
                     <archive>
                         <manifest>
                             <addClasspath>true</addClasspath>

--- a/native-maven-plugin/src/testFixtures/groovy/org/graalvm/buildtools/maven/AbstractGraalVMMavenFunctionalTest.groovy
+++ b/native-maven-plugin/src/testFixtures/groovy/org/graalvm/buildtools/maven/AbstractGraalVMMavenFunctionalTest.groovy
@@ -139,24 +139,61 @@ abstract class AbstractGraalVMMavenFunctionalTest extends Specification {
     }
 
     void mvn(List<String> args) {
-        mvn(args as String[])
+        Map<String, String> systemProperties = [
+            "org.apache.http" : "off",
+            "org.apache.http.wire" : "off",
+        ].collectEntries { key, value ->
+            ["org.slf4j.simpleLogger.log.${key}".toString(), value]
+        }
+        mvn(args, systemProperties)
     }
 
     void mvn(String... args) {
-        System.out.println("Running copy of maven project `" + testOrigin + "` with `" + args + "`");
+        mvn(args.toList())
+    }
+
+    void mvn(List<String> args, Map<String, String> systemProperties) {
+        System.out.println("Running copy of maven project `" + testOrigin + "` with " + args);
+        var resultingSystemProperties = [
+                "common.repo.uri": System.getProperty("common.repo.uri"),
+                "seed.repo.uri": System.getProperty("seed.repo.uri"),
+                "maven.repo.local": testDirectory.resolve("local-repo").toFile().absolutePath
+        ]
+        resultingSystemProperties.putAll(systemProperties)
+
         result = executor.execute(
                 testDirectory.toFile(),
-                [
-                        "org.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener": "warn",
-                        "common.repo.uri": System.getProperty("common.repo.uri"),
-                        "seed.repo.uri": System.getProperty("seed.repo.uri"),
-                        "maven.repo.local": testDirectory.resolve("local-repo").toFile().absolutePath
-                ],
+                resultingSystemProperties,
                 [*injectedSystemProperties,
                  *args],
                 new File(System.getProperty("maven.settings"))
         )
         System.out.println("Exit code is ${result.exitCode}")
+
+    }
+
+    void mvnDebug(String... args) {
+        Map<String, String> systemProperties = [
+            "org.apache.http"                                                    : "off",
+            "org.apache.http.wire"                                               : "off",
+            "org.apache.maven.cli.transfer.Slf4jMavenTransferListener"           : "warn",
+            "org.eclipse.aether.internal.impl.DefaultTransporterProvider"        : "error",
+            "org.eclipse.aether.internal.impl.DefaultRepositoryConnectorProvider": "error",
+            "org.eclipse.aether.internal.impl.TrackingFileManager"               : "error",
+            "org.eclipse.aether.internal.impl.DefaultArtifactResolver"           : "error",
+            "org.codehaus.plexus.PlexusContainer"                                : "error",
+            "org.apache.maven.plugin.internal.DefaultPluginDependenciesResolver" : "error",
+            "org.apache.maven.shared.filtering.DefaultMavenFileFilter"           : "error",
+            "org.codehaus.mojo.exec.ExecMojo"                                    : "debug",
+            // This is THE logger we need for checking java execution
+            "org.apache.maven.lifecycle.internal.LifecycleDebugLogger"           : "error"
+        ].collectEntries { key, value ->
+            ["org.slf4j.simpleLogger.log.${key}".toString(), value]
+        }
+
+        ArrayList<String> resultingArgs = args.toList()
+        resultingArgs.add(0, "-X")
+        mvn(resultingArgs, systemProperties)
     }
 
     boolean isDidNotCrash() {

